Add StatefulSet rollback and rolling update test with PVCs

jarrpa opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Adds a StatefulSet e2e test for testing rolling upgrades and rollbacks on a StatefulSet using PVCs.

**Which issue(s) this PR fixes**:
Fixes #63940

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```
NONE
```

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

jarrpa committed:
Add StatefulSet rollback and rolling update test with PVCs

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

krmayankk commented:
/ok-to-test

krmayankk commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,123 @@ var _ = SIGDescribe("StatefulSet", func() {
+			Expect(priorRevision).To(Equal(updateRevision),
+				"Prior revision should equal update revision during roll back")
+
+			By("Rolling back update in reverse ordinal order")
```
How are you testing rollback ? Statefulsets doesn’t support it or did you mean rolling update ?

jarrpa replied:
This is mostly a copy of the test starting [here](https://github.com/kubernetes/kubernetes/pull/64350/files/cf7ddbe837d2baf1918230c641672c4e6ae70134#diff-cb00e9492a5e88ff0e10e4b82878d690R373), so I'm just doing what they did. Looks like they update the container image to a previous version and call that a rollback.

jarrpa commented:
/test pull-kubernetes-node-e2e

kargakis commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,123 @@ var _ = SIGDescribe("StatefulSet", func() {
+			}
+
+			By("Rolling back to a previous revision")
+			sst.BreakPodHttpProbe(ss, &pods.Items[1])
```
err = sst.BreakPodHttpProbe(ss, &pods.Items[1])

kargakis commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,123 @@ var _ = SIGDescribe("StatefulSet", func() {
+			By("Updating Pods in reverse ordinal order")
+			pods = sst.GetPodList(ss)
+			sst.SortStatefulPods(pods)
+			sst.RestorePodHttpProbe(ss, &pods.Items[1])
```
check the returned error

kargakis commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,123 @@ var _ = SIGDescribe("StatefulSet", func() {
+						currentRevision))
+			}
+			sst.SortStatefulPods(pods)
+			sst.BreakPodHttpProbe(ss, &pods.Items[1])
```
check the returned error

jarrpa commented:
@kargakis Updated as requested. I'll note that I took this code from another test below, so I imagine there should be another PR to fix that as well.

jarrpa commented:
/test pull-kubernetes-verify

foxish commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,124 @@ var _ = SIGDescribe("StatefulSet", func() {
+						currentRevision))
+			}
+			sst.SortStatefulPods(pods)
+			err = sst.BreakPodHttpProbe(ss, &pods.Items[1])
```
What's the purpose of breaking and restoring the http probe in the context of this test?

jarrpa replied:
No idea, this is taken from the very similar test below so I figured there must have been some reason for it.

msau42 replied:
Since you're reusing the existing StatefulSet test but with PVCs, maybe it would be better to wrap the existing test in a common function that could be called by both tests.

jarrpa replied:
Sorry for the delay. Sure thing, testing it now. We'll see how bad my implementation is. :)

foxish commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,6 +248,124 @@ var _ = SIGDescribe("StatefulSet", func() {
 
+		// This can't be Conformance yet because it depends on a default
+		// StorageClass and a dynamic provisioner.
+		It("should perform rolling updates and roll backs of template modifications with PVCs", func() {
```
From the test description, I'd have expected that we'd exercise some PVC option explicitly. I think the test name could use revising.

foxish replied:
Hmm, from the linked issue, does this actually use an RWO PVC?

jarrpa commented:
Rebased and updated the patch per @msau42's suggestion to collapse most of the two tests into a common function. I've run the tests successfully in my local environment using `local-up-cluster.sh` with the host-path provisioner.

jarrpa commented:
/retest

msau42 commented on test/e2e/apps/statefulset.go:
```diff
@@ -248,14 +248,8 @@ var _ = SIGDescribe("StatefulSet", func() {
-			By("Creating a new StatefulSet")
-			ss := framework.NewStatefulSet("ss2", ns, headlessSvcName, 3, nil, nil, labels)
+		// This function is used by the following two tests.
+		rollbackTest := func(ss *apps.StatefulSet) {
```
Instead of being an inline function, can this just be a named function?

jarrpa replied:
Can do, should it just go above this current function?

jarrpa replied:
Maybe. It's referencing variables instantiated within the enclosing function, so there would need to be several parameters or a new struct, right? Also, where would it go in the file, ideally?

msau42 replied:
In other test files, I've seen the test functions just go at the bottom

msau42 reviewed (commented):
lgtm, just a minor comment

jarrpa commented:
@msau42 Updated. Moved common code into a function at the bottom of the file. Tests pass on my local machine.

jarrpa commented:
Failing test has failed twice on:
```
[sig-api-machinery] Garbage collector should not be blocked by dependency circle [Conformance]
```

msau42 commented:
/lgtm

jarrpa commented:
/assign @kargakis

kargakis commented:
/approve
